tests/executables_src/testDoubleBuffering.cpp

```diff
@@ -160,10 +219,263 @@ template<typename Register>
 uint32_t AreaType<Register>::_currentBufferNumber = 0;
 
 BOOST_AUTO_TEST_CASE(testUnified) {
-  std::string lmap = "(logicalNameMap?map=doubleBuffer.xlmap&target=" + rawDeviceCdd + ")";
   ChimeraTK::UnifiedBackendTest<>().addRegister<AreaType<MyArea1>>().runTests(lmap);
 }
 
+struct DeviceFixture {
+  Device d;
+  boost::shared_ptr<NDRegisterAccessor<uint32_t>> doubleBufferingEnabled;
+  // we call the backend frontdoor when we modify the behavior of the thread which reads via double buffering mechanism
+  boost::shared_ptr<DummyForDoubleBuffering> frontdoor;
+  DeviceFixture() : d(lmap) {
+    // before any access, also via backdoor, must open
+    d.open();
+    frontdoor = boost::dynamic_pointer_cast<DummyForDoubleBuffering>(backdoor);
+    assert(frontdoor);
+    doubleBufferingEnabled = backdoor->getRegisterAccessor<uint32_t>("APP/1/WORD_DUB_BUF_ENA", 0, 0, {});
+    doubleBufferingEnabled->accessData(0) = 1;
+    doubleBufferingEnabled->write();
+  }
+};
+
+BOOST_FIXTURE_TEST_CASE(testSlowReader, DeviceFixture) {
+  /*
+   * Test race condition: slow reader, which blocks the Firmware from buffer switching.
+   */
+  auto accessor = d.getOneDRegisterAccessor<uint32_t>("/doubleBuffer");
+
+  // make double buffer operation block after write to ctrl register, at read of buffer number
+  std::thread s([&] {
+    // this thread reads from double-buffered region
+    frontdoor->blockNextRead[0] = true;
+    accessor.read();
+  });
+
+  // wait that thread s is in blocked double-buffer read
+  frontdoor->blockedInRead[0].wait();
+
+  // simplification: instead of writing fw simulation which would overwrite data now,
+  // just check that buffer switching was disabled
+  doubleBufferingEnabled->readLatest();
+  BOOST_CHECK(!doubleBufferingEnabled->accessData(0));
+
+  frontdoor->unblockRead[0].wait();
+  s.join();
+
+  // check that buffer switching enabled, by finalization of double-buffered read
+  doubleBufferingEnabled->readLatest();
+  BOOST_CHECK(doubleBufferingEnabled->accessData(0));
+}
+
+BOOST_FIXTURE_TEST_CASE(testConcurrentRead, DeviceFixture) {
+  /*
+   * A test which exposes the dangerous race condition of two readers
```

Comment:
Actually, I am not so sure any more that it is easy to trigger the remaining race condition. This test is checking the main purpose (two accessors starting read quasi simultaneously, and after the first one is finished the double buffering is still disabled).

Reply:
To trigger above-mentioned race condition, you need to start another reader, just after the first reader counted down so readerCount=0 but before it wrote to the firmware control register (to enable buffer switching again). That's hard to get in a test.
Interesting fact: if the firmware control register would already provide the atomic readerCount variable (instead of buffer switch enable register), we would not have the race condition at all.

Reply:
A test that would be sensitive to this race condition would block forever. It is not easy to implement. The code review is sufficient here.